### ISSUE TITLE
Potential exploit involving large volumes of node metadata

### ISSUE THREAD

**Issue by sorcerykid** [luanti-org/luanti#8182]
##### Issue type
 - Bug report
 
##### Summary
A potential exploit has been recently discovered in Minetest Game as described in https://github.com/minetest/minetest_game/pull/2302

Consider the worst case scenario: A mapblock can theoretically contain up to 2048 signs (assuming interleaving walls with signs attached). That means potentially up to 1,048,576 bytes of user text being stored in a single mapblock blob. Now, let's go one step further and make these walls of chests, with full inventory of written books. (I'm not sure if there is a text limit on books, but let's go with 512 bytes, which is extremely conservative).

This equates to

- 2048 chests / mapblock
- 8 x 4 written books / chest
- 512 bytes of text / per written book

For a grand total (in addition to the signs above) of _34.6 megabytes_ of user text. And, this assumes that written books are limited to 512 bytes, which I doubt. Of course estimate doesn't even account for the overhead of serialization of item metadata. I think there needs to be some type of safeguards (within the engine) to protect against the intentional bloating of mapblocks by users.

##### Solutions/Workarounds
I would like to open a discussion on both long-term solutions as well as immediate workarounds. From a design standpoint, some possible candidates for engine changes include
* storing node metadata in a separately indexed table, 
* transmitting node metadata to clients asynchronously, or
* imposing hard limits on the size of blobs in the database

I recognize that, all three proposals would require dramatic changes to the legacy database format. But perhaps this is the impetus to develop a more scalable and robust map storage architecture for Minetest 5.x. There has been ongoing discussion of layered maps, partitioned maps, etc. as well as the desire for higher content IDs, extended params, etc. all of which is hindered by the existing BLOB format.

I'd be interested in hearing people's feedback. Is this a vulnerability, or am I missing something. And if so, what measures can be taken in the interim to mitigate against such threats?

**Comment by sofar** [comment]
Most servers run the risk of going out of disk space unless they limit user interactions to only a small portion of the map due to mapgen creating large amounts of block data, even without metadata in consideration.

The point being that limiting sizes of "things" on the server is entirely reasonable and should be applied on a per-server basis, and multiple strategies should be supported. One could for instance argue that a limitation per player is likely going to be far more effective than a limit per mapblock, which can be shared by several players, or even many.

From that perspective, I would think that if the engine would set hard limits, that they are based on functionality, and not nuisance levels. In other words, the engine should set limits that, if not enforced, would cause failure. The game should set limits at much lower levels to prevent abuse.

For any limits to be adopted by the engine, I would hope that actual data is used to determine what those levels are, and that design changes take into account whether those levels aren't much lower because of more complex storage requirements, after all, if you split out metadata into different storage, it would take more database trips to fetch the needed data, and so there's a potential risk for a performance loss, further making performance a potential concern.

There are things you can do without engine modifications. Quota systems are widely used and one can attribute all metadata changes in the map to specific users, and that could very well be a simple way to deal with players that exceed a hypothetical quota. Not cheap, of course, but certainly doable.

As an aside, the engine already limits entities per mapblock. I can envision a limit on metadata, although I have no idea how I would handle an overflow. The current metadata API does not handle errors at all.

**Comment by random-geek** [comment]
I can't currently think of any reliable way to prevent this from happening without putting players' possessions at risk.

Fortunately, all node data is compressed, and compression algorithms take advantage of repetition to reduce the number of bytes needed to store a particular block of data. Copying and pasting some text into books over and over again, which is the most obvious way to generate a lot of raw metadata, is simply not going to work.
Given this, someone particularly evil may decide to write a script to generate and paste 10 KB of random data into a book. In this scenario, the effect of data compression would be minimal. If each book contained about 10 KB of data, then each chest could "hold" 320 KB. However, pasting all this data would be time-consuming and the player would hopefully give up after 5 or 10 chests.
At this point, the mapblock might be 4 MB or so, which would not affect the size of the database too much; however, it would take a while to send to clients, especially ones with slower connections. You might think that a player could automate the book-writing and chest-placing process as well, but having tried to do things similar to this in the past, it doesn't really work as soon as there is any amount of lag. 
The issue is certainly a valid one, but I feel like this method of cheating would be too hard and time-consuming for most aspiring hackers. If I was this evil, I would rather spend my time looking for less obvious ways to bring down a server, such as SQL injections, etc. :-)

**Comment by Fixer-007** [comment]
In some ways this is a problem in minecraft too, but I think they have 1 megabyte limit or so for 16x256x16 chunk, there was dupe exploit related to exceeding this limit.

**Comment by rubenwardy** [comment]
The correct approach to this is to apply not one but multiple different methods of rate limiting

**Comment by erlehmann** [comment]
> In some ways this is a problem in minecraft too, but I think they have 1 megabyte limit or so for 16x256x16 chunk, there was dupe exploit related to exceeding this limit.

Yes. The problem Minecraft encountered is that the naive solution to just not accept any changes to a mapblock if it is ”full” means that you can take stuff out of it, then stuff it with garbage and have it reset to whatever it was previously.

**Comment by sfan5** [comment]
closing in favor of #13365, which has more info IMO.
